<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>MA-plots using ggplot2 — ggjammaplot • jamma</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="MA-plots using ggplot2 — ggjammaplot"><meta property="og:description" content="MA-plots using ggplot2"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jamma</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.21.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/jamma-vignette.html">Introduction to jamma</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jmw86069/jamma/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MA-plots using ggplot2</h1>
    <small class="dont-index">Source: <a href="https://github.com/jmw86069/jamma/blob/HEAD/R/jamma-ggjammaplot.R" class="external-link"><code>R/jamma-ggjammaplot.R</code></a></small>
    <div class="hidden name"><code>ggjammaplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>MA-plots using ggplot2</p>
    </div>

    <div class="ref-usage sourceCode"><pre class="sourceCode r"><code><span class="fu">ggjammaplot</span><span class="op">(</span>
  <span class="va">x</span>,
  nbin_factor <span class="op">=</span> <span class="fl">1</span>,
  bw_factor <span class="op">=</span> <span class="fl">1</span>,
  assay_name <span class="op">=</span> <span class="fl">1</span>,
  useMedian <span class="op">=</span> <span class="cn">FALSE</span>,
  controlSamples <span class="op">=</span> <span class="cn">NULL</span>,
  centerGroups <span class="op">=</span> <span class="cn">NULL</span>,
  colramp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"transparent"</span>, <span class="st">"lightblue"</span>, <span class="st">"blue"</span>, <span class="st">"navy"</span>, <span class="st">"orange"</span>, <span class="st">"orangered2"</span><span class="op">)</span>,
  groupedX <span class="op">=</span> <span class="cn">TRUE</span>,
  grouped_mad <span class="op">=</span> <span class="cn">TRUE</span>,
  outlierMAD <span class="op">=</span> <span class="fl">5</span>,
  mad_row_min <span class="op">=</span> <span class="fl">4</span>,
  displayMAD <span class="op">=</span> <span class="cn">FALSE</span>,
  noise_floor <span class="op">=</span> <span class="fl">0</span>,
  noise_floor_value <span class="op">=</span> <span class="cn">NA</span>,
  naValue <span class="op">=</span> <span class="cn">NA</span>,
  centerFunc <span class="op">=</span> <span class="va">centerGeneData</span>,
  whichSamples <span class="op">=</span> <span class="cn">NULL</span>,
  useRank <span class="op">=</span> <span class="cn">FALSE</span>,
  titleBoxColor <span class="op">=</span> <span class="st">"lightgoldenrod1"</span>,
  outlierColor <span class="op">=</span> <span class="st">"lemonchiffon"</span>,
  fillBackground <span class="op">=</span> <span class="cn">TRUE</span>,
  maintitle <span class="op">=</span> <span class="cn">NULL</span>,
  subtitle <span class="op">=</span> <span class="cn">NULL</span>,
  summary <span class="op">=</span> <span class="st">"mean"</span>,
  difference <span class="op">=</span> <span class="st">"difference"</span>,
  transFactor <span class="op">=</span> <span class="fl">0.2</span>,
  doPlot <span class="op">=</span> <span class="cn">TRUE</span>,
  highlightPoints <span class="op">=</span> <span class="cn">NULL</span>,
  highlightPch <span class="op">=</span> <span class="fl">21</span>,
  highlightCex <span class="op">=</span> <span class="fl">1.5</span>,
  highlightColor <span class="op">=</span> <span class="cn">NULL</span>,
  doHighlightLegend <span class="op">=</span> <span class="cn">TRUE</span>,
  ablineH <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span>,
  base_size <span class="op">=</span> <span class="fl">12</span>,
  panel.grid.major.colour <span class="op">=</span> <span class="st">"grey90"</span>,
  panel.grid.minor.colour <span class="op">=</span> <span class="st">"grey95"</span>,
  return_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ggplot"</span>, <span class="st">"data"</span><span class="op">)</span>,
  xlim <span class="op">=</span> <span class="cn">NULL</span>,
  ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">6</span>, <span class="fl">6</span><span class="op">)</span>,
  ncol <span class="op">=</span> <span class="cn">NULL</span>,
  nrow <span class="op">=</span> <span class="cn">NULL</span>,
  blankPlotPos <span class="op">=</span> <span class="cn">NULL</span>,
  verbose <span class="op">=</span> <span class="cn">FALSE</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>

    <div id="ref-sections">
    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>one of the following inputs:</p><ul><li><p><code>numeric</code> matrix</p></li>
<li><p><code>SummarizedExperiment</code> object, where the
assay data is defined using <code>assays(x)[[assay_name]]</code>. Accordingly,
<code>assay_name</code> can be either an integer index or <code>character</code> string
matching the <code>names(assays(x))</code>.</p></li>
<li><p><code>list</code> output from <code><a href="jammacalc.html">jammacalc()</a></code> or <code><a href="jammaplot.html">jammaplot()</a></code>, where each
element in the list is a two-column <code>matrix</code> with colnames <code>c("x", "y")</code>.</p></li>
</ul></dd>
<dt>nbin_factor</dt>
<dd><p><code>numeric</code> value used to adjust the number of bins
used to display the MA-plots, where values higher than <code>1</code> increase
the resolution and level of detail, and values below <code>1</code> decrease
the resolution. Note the number of bins are already adjusted based
upon the square root of the number of plot panels, and <code>nbin_factor</code>
applied to that value.</p></dd>
<dt>bw_factor</dt>
<dd><p><code>numeric</code> used to adjust the resolution of the
2-dimensional bandwidth calculation, where higher values create
more detailed density, and lower values create a smoother density
across the range of data.</p></dd>
<dt>assay_name</dt>
<dd><p>relevant only when <code>x</code> is <code>SummarizedExperiment</code>,
one of these input types:</p><ul><li><p><code>character</code> string that matches <code>names(assays(x))</code></p></li>
<li><p><code>integer</code> index for <code>assays(x)</code>, where any value higher than
<code>length(assays(x))</code> is adjusted to <code>length(assays(x))</code>, which makes
it convenient to select the last element in the list of <code>assays(x)</code>
by using <code>assay_name = Inf</code>.</p></li>
</ul></dd>
<dt>useMedian</dt>
<dd><p><code>logical</code> indicating whether calculations should use
<code>median</code>, or when <code>useMedian=FALSE</code> the <code>mean</code> is used. The median
has the benefit of reducing effect of outliers, however the mean
has the advantage that it represents data consistent with most
parametric statistical analyses.</p></dd>
<dt>controlSamples</dt>
<dd><p><code>character</code> vector of <code>colnames(x)</code> to use as
the control when calculating centered data. By default, all samples
are used, so the classic MA-plot is the value of each sample,
subtracting the median or mean value calculated across all samples.
It is sometimes useful to define a subset of known samples for this
calculation, which can be beneficial in avoiding outliers, or for
consistency by selecting high quality control samples.</p></dd>
<dt>centerGroups</dt>
<dd><p><code>character</code> vector with length equal to <code>ncol(x)</code>,
which defines subgroups of <code>colnames(x)</code> to be treated independently
during the MA-plot calculation.</p></dd>
<dt>colramp</dt>
<dd><p>one of several inputs recognized by
<code><a href="https://rdrr.io/pkg/jamba/man/getColorRamp.html" class="external-link">jamba::getColorRamp()</a></code>. It typically recognizes either the name of
a color ramp from RColorBrewer, the name of functions from the
<code>viridis</code> package such as <code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis::viridis()</a></code>, or single R colors, or
a vector of R colors. When a single color is supplied, a gradient
is created from white to that color, where the default base color
can be customized with <code>defaultBaseColor="black"</code> for example.</p></dd>
<dt>groupedX</dt>
<dd><p><code>logical</code> indicating whether the x-axis value, which
represents the median or mean value, should be calculated independently
for each group when <code>centerGroups</code> is used with multiple groups.
Typically <code>groupedX=TRUE</code> is recommended, however it can be beneficial
to share an overall x-axis value in specific circumstances.</p></dd>
<dt>grouped_mad</dt>
<dd><p><code>logical</code> indicating whether the MAD factor calculation
of variability among samples should be performed independently
for each group when <code>centerGroups</code> is used with multiple groups.
Typically <code>grouped_max=TRUE</code> is recommended, however it can be beneficial
to share an overall MAD factor threshold across all samples
in specific circumstances.</p></dd>
<dt>outlierMAD</dt>
<dd><p><code>numeric</code> indicating the MAD factor threshold above
which a particular sample is considered an outlier.</p></dd>
<dt>mad_row_min</dt>
<dd><p><code>numeric</code> value indicating the minimum x-axis
value, calculated using either median or mean as defined by
argument <code>useMedian</code>, at or above which a measurement is used in the
MAD factor calculation. This threshold is useful to restrict the
MAD variability calculation to measurements (rows in <code>x</code>) with
signal that meets a minimum noise threshold.</p></dd>
<dt>displayMAD</dt>
<dd><p><code>logical</code> indicating whether to display the MAD factor
in the bottom right corner of each MA-plot panel.</p></dd>
<dt>noise_floor, noise_floor_value</dt>
<dd><p><code>numeric</code> to define a numeric
floor, or <code>NULL</code> for no numeric floor. Values <strong>at or below</strong>
<code>noise_floor</code> are set to <code>noise_floor_value</code>, intended for two
potential uses:</p><ol><li><p>Filter out value below a threshold, so they do not affect centering.</p><ul><li><p>This option is valuable to remove zeros when a zero <code>0</code> is considered
"no measurement observed", typically for count data such as RNA-seq,
NanoString, and especially single-cell protocols or other protocols
that produce a large number of missing values.</p></li>
<li><p>One can typically tell whether input data includes zero <code>0</code>
values by the presence of characteristic 45-degree angle lines
originating from <code>x=0</code> angled toward the right. The points along
this line are rows with more measurements of zero than non-zero,
there this sample has a non-zero value.</p></li>
</ul></li>
<li><p>Set values at a noise floor to the noise floor, to retain the
measurement but minimize the effect during centering to the lowest
realiable measurement for the platform technology.</p><ul><li><p>This value may be set to a platform noise floor
for something like microarray data where the intensity may be
unreliable below a threshold; or</p></li>
<li><p>for quantitative PCR measurements where cycle threshold (Ct)
values may become unreliable, for example above CT=40 or CT=35.
Data is often transformed to abundance with <code>2 ^ (40 - CT)</code> then
log2-transformed for analysis. In this case, to apply a <code>noise_floor</code>
effective for CT=35, one would use <code>noise_floor=5</code>.</p></li>
</ul></li>
</ol></dd>
<dt>naValue</dt>
<dd><p><code>character</code> string used to convert values of <code>NA</code> to
something else. This argument is useful when a numeric matrix may
contain <code>NA</code> values but would prefer them to be, for example, <code>0</code>.</p></dd>
<dt>centerFunc</dt>
<dd><p><code>function</code> used to supply a custom data centering
function. In practice this argument should rarely be changed.</p></dd>
<dt>whichSamples</dt>
<dd><p><code>integer</code> index of samples in <code>colnames(x)</code> to be
plotted, however all samples in <code>colnames(x)</code> will be used for the
MA-plot calculations and data centering. This argument is intended
to help zoom in to inspect a specific subset of samples, without
having to plot all samples in <code>x</code>.</p></dd>
<dt>useRank</dt>
<dd><p><code>logical</code> indicating whether to plot rank on the x-axis,
rank-difference on the y-axis for each sample. This transformation
is rather useful, especially when downstream analysis tools may
also refer to the rank value of particular measurements.</p></dd>
<dt>titleBoxColor</dt>
<dd><p><code>character</code> vector of R colors, where
<code>titleBoxColor</code> is equal to <code>ncol(x)</code>, or where
<code>names(titleBoxColor)</code> matches <code>colnames(x)</code>. When supplied, each
plot panel strip background will be colored accordingly.</p></dd>
<dt>outlierColor</dt>
<dd><p><code>character</code> string representing one R color,
used when <code>colrampOutlier</code> is <code>NULL</code> and when <code>outlierMAD</code> is
defined. This color is used for MA-plot outlier panels by
substituting the first color from the <code>colramp</code> color ramp,
to act as a visual cue that the panel represents an outlier.</p></dd>
<dt>fillBackground</dt>
<dd><p><code>logical</code> currently used for base R graphics
output, and passed to <code><a href="https://rdrr.io/pkg/jamba/man/plotSmoothScatter.html" class="external-link">jamba::plotSmoothScatter()</a></code>,
indicating whether to fill the plot panel using the
first color in the color ramp for each MA-plot panel, or when
a plot panel is an outlier, it uses <code>outlierColor</code>.
This argument is mainly useful to highlight outlier panels,
although it is also useful when the color ramp has non-white
base color, for example <code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis::viridis()</a></code>.</p></dd>
<dt>maintitle</dt>
<dd><p><code>character</code> string with the title displayed above
all individual MA-plot panels. It will appear in the top outer
margin.</p></dd>
<dt>subtitle</dt>
<dd><p><code>NULL</code> or <code>character</code> vector to be drawn at
the bottom left corner of each plot panel, the location
is defined by <code>subtitlePreset</code>.</p></dd>
<dt>transFactor</dt>
<dd><p><code>numeric</code> adjustment to the visual density of
smooth scatter points. For base R graphics, this argument is
passed to <code><a href="https://rdrr.io/pkg/jamba/man/plotSmoothScatter.html" class="external-link">jamba::plotSmoothScatter()</a></code>. The argument value is based upon
<code><a href="https://rdrr.io/r/graphics/smoothScatter.html" class="external-link">graphics::smoothScatter()</a></code> argument <code>transformation</code>, which uses
default <code>function(x)x^0.25</code>. The <code>transFactor</code> is equivalent to the
exponential in the form: <code>function(x)x^transFactor</code>. Lower values
make the point density more visually intense, higher values make the
point density less visually intense.</p></dd>
<dt>doPlot</dt>
<dd><p><code>logical</code> indicating whether to create plots. When
<code>doPlot=FALSE</code> only the MA-plot panel data is returned.</p></dd>
<dt>highlightPoints</dt>
<dd><p><code>NULL</code>, or <code>character</code> vector, or a
<code>list</code> of <code>character</code> vectors indicating <code>rownames(x)</code> to
highlight in each MA-plot panel. When <code>NULL</code>, no points are
highlighted; when <code>character</code> vector, points are highlighted in
all MA-plot panels; when <code>list</code> of <code>character</code> vectors, each
<code>character</code> vector in the list is highlighted using a unique
color in <code>highlightColor</code>. Points are drawn using
<code><a href="https://rdrr.io/r/graphics/points.html" class="external-link">graphics::points()</a></code> and colored using <code>highlightColor</code>,
which can be time-consuming for a large number of highlight
points.</p></dd>
<dt>highlightCex</dt>
<dd><p><code>numeric</code> value recycled to <code>length(highlightPoints)</code>
indicating the highlight point size.</p></dd>
<dt>highlightColor</dt>
<dd><p><code>character</code> vector used when <code>highlightPoints</code>
is defined. It is recycled to <code>length(highlightPoints)</code> and
is applied either to</p></dd>
<dt>doHighlightLegend</dt>
<dd><p><code>logical</code> indicating whether to print a
color legend when <code>highlightPoints</code> is defined. The legend is
displayed in the bottom outer margin of the page using
<code><a href="outer_legend.html">outer_legend()</a></code>, and the page is adjusted to add bottom
outer margin.</p></dd>
<dt>ablineH</dt>
<dd><p><code>numeric</code> vector indicating position of
horizontal and vertical lines in each MA-plot panel.</p></dd>
<dt>xlim</dt>
<dd><p><code>NULL</code> or <code>numeric</code> vector <code>length=2</code> indicating
the y-axis and x-axis ranges, respectively. The values are useful
to define consistent dimensions across all panels. The
default <code>ylim=c(-4,4)</code> represents 16-fold up and down range in
normal space, and is typically a reasonable starting point for
most purposes. Even if numeric values are all between
<code>-1.5</code> and <code>1.5</code>, it is still recommended to keep a range in
context of <code>c(-4, 4)</code>, to indicate that the observed values
are lower than typically observed. The <code>c(-4, 4)</code> may be adjusted
relative to the typical ranges expected for the data.
It is sometimes helpful to define <code>xlim</code> slightly above zero for
datasets that have an extremely large proportion of zeros, in order
to reduce the visual effect of having that much point density at
zero, for example with <code>xlim=c(0.001, 20)</code> and
<code>applyRangeCeiling=FALSE</code>.</p></dd>
<dt>ylim</dt>
<dd><p><code>NULL</code> or <code>numeric</code> vector <code>length=2</code> indicating
the y-axis and x-axis ranges, respectively. The values are useful
to define consistent dimensions across all panels. The
default <code>ylim=c(-4,4)</code> represents 16-fold up and down range in
normal space, and is typically a reasonable starting point for
most purposes. Even if numeric values are all between
<code>-1.5</code> and <code>1.5</code>, it is still recommended to keep a range in
context of <code>c(-4, 4)</code>, to indicate that the observed values
are lower than typically observed. The <code>c(-4, 4)</code> may be adjusted
relative to the typical ranges expected for the data.
It is sometimes helpful to define <code>xlim</code> slightly above zero for
datasets that have an extremely large proportion of zeros, in order
to reduce the visual effect of having that much point density at
zero, for example with <code>xlim=c(0.001, 20)</code> and
<code>applyRangeCeiling=FALSE</code>.</p></dd>
<dt>ncol</dt>
<dd><p><code>integer</code> number of MA-plot panel columns and rows
passed to <code>graphics::par("mfrow")</code> when <code>doPar=TRUE</code>. When only one
value is supplied, <code>nrow</code> or <code>ncol</code>, the other value is defined
by <code>ncol(x)</code> and <code>blankPlotPos</code> so all panels can be contained on
one page. When <code>nrow</code> and <code>ncol</code> are defined such that multiple
pages are produced, each page will be annotated with <code>maintitle</code>
and <code>doHighlightLegend</code> as relevant.</p></dd>
<dt>nrow</dt>
<dd><p><code>integer</code> number of MA-plot panel columns and rows
passed to <code>graphics::par("mfrow")</code> when <code>doPar=TRUE</code>. When only one
value is supplied, <code>nrow</code> or <code>ncol</code>, the other value is defined
by <code>ncol(x)</code> and <code>blankPlotPos</code> so all panels can be contained on
one page. When <code>nrow</code> and <code>ncol</code> are defined such that multiple
pages are produced, each page will be annotated with <code>maintitle</code>
and <code>doHighlightLegend</code> as relevant.</p></dd>
<dt>blankPlotPos</dt>
<dd><p><code>NULL</code> or <code>integer</code> vector indicating
plot panel positions to be drawn blank, and therefore skipped.
Plot panels are drawn in the exact order of <code>colnames(x)</code> received.
Blank panel positions are intended to help customize the visual
alignment of MA-plot panels. The mechanism is similar to
<code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">ggplot2::facet_wrap()</a></code> except that blank positions can be manually
defined by what makes sense for the experiment design.</p></dd>
<dt>verbose</dt>
<dd><p>logical indicating whether to print verbose output.</p></dd>
<dt>...</dt>
<dd><p>additional parameters sent to downstream functions,
<code><a href="https://rdrr.io/pkg/jamba/man/plotSmoothScatter.html" class="external-link">jamba::plotSmoothScatter</a></code>, <code><a href="centerGeneData.html">centerGeneData</a></code>.</p></dd>
</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>This method is under active development and may change as
features are implemented.</p>
<p>It is currently fully functional and is being documented.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other jam plot functions: 
<code><a href="jammaplot.html">jammaplot</a>()</code></p></div>
    </div>
    </div>

    <div id="examples">
    <h2>Examples</h2>
    <div class="ref-examples sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jamba/man/check_pkg_installed.html" class="external-link">check_pkg_installed</a></span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span>
<span class="r-in">   <span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jamba/man/check_pkg_installed.html" class="external-link">check_pkg_installed</a></span><span class="op">(</span><span class="st">"farrisdata"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in">   <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span><span class="op">)</span>;</span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">GeneSE</span> <span class="op">&lt;-</span> <span class="fu">farrisdata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/farrisdata/man/farrisGeneSE.html" class="external-link">farrisGeneSE</a></span>;</span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">titleBoxColor</span> <span class="op">&lt;-</span> <span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jamba/man/nameVector.html" class="external-link">nameVector</a></span><span class="op">(</span></span>
<span class="r-in">      <span class="fu">farrisdata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/farrisdata/man/colorSub.html" class="external-link">colorSub</a></span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">GeneSE</span><span class="op">)</span><span class="op">$</span><span class="va">groupName</span><span class="op">)</span><span class="op">]</span>,</span>
<span class="r-in">      <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">GeneSE</span><span class="op">)</span><span class="op">)</span>;</span>
<span class="r-in">   <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"warn"</span><span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>;</span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">gg</span> <span class="op">&lt;-</span> <span class="fu">ggjammaplot</span><span class="op">(</span><span class="va">GeneSE</span>,</span>
<span class="r-in">      ncol<span class="op">=</span><span class="fl">6</span>,</span>
<span class="r-in">      base_size<span class="op">=</span><span class="fl">12</span>,</span>
<span class="r-in">      assay_name<span class="op">=</span><span class="st">"raw_counts"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">gg</span> <span class="op">&lt;-</span> <span class="fu">ggjammaplot</span><span class="op">(</span><span class="va">GeneSE</span>,</span>
<span class="r-in">      ncol<span class="op">=</span><span class="fl">6</span>,</span>
<span class="r-in">      assay_name<span class="op">=</span><span class="st">"counts"</span>,</span>
<span class="r-in">      useRank<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span class="r-in">      ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">11000</span>, <span class="fl">11000</span><span class="op">)</span>,</span>
<span class="r-in">      maintitle<span class="op">=</span><span class="st">"MA-plots by rank and rank difference"</span>,</span>
<span class="r-in">      titleBoxColor<span class="op">=</span><span class="va">titleBoxColor</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">gg</span> <span class="op">&lt;-</span> <span class="fu">ggjammaplot</span><span class="op">(</span><span class="va">GeneSE</span>,</span>
<span class="r-in">      ncol<span class="op">=</span><span class="fl">6</span>,</span>
<span class="r-in">      assay_name<span class="op">=</span><span class="st">"counts"</span>,</span>
<span class="r-in">      titleBoxColor<span class="op">=</span><span class="va">titleBoxColor</span>,</span>
<span class="r-in">      base_size<span class="op">=</span><span class="fl">10</span>,</span>
<span class="r-in">      maintitle<span class="op">=</span><span class="st">"MA-plots showing MAD factor"</span>,</span>
<span class="r-in">      displayMAD<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">gg</span> <span class="op">&lt;-</span> <span class="fu">ggjammaplot</span><span class="op">(</span><span class="va">GeneSE</span>,</span>
<span class="r-in">      ncol<span class="op">=</span><span class="fl">6</span>,</span>
<span class="r-in">      assay_name<span class="op">=</span><span class="st">"counts"</span>,</span>
<span class="r-in">      titleBoxColor<span class="op">=</span><span class="va">titleBoxColor</span>,</span>
<span class="r-in">      maintitle<span class="op">=</span><span class="st">"MA-plot omitting one panel, then using blankPlotPos"</span>,</span>
<span class="r-in">      whichSamples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">GeneSE</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">21</span>, <span class="fl">23</span><span class="op">:</span><span class="fl">24</span><span class="op">)</span><span class="op">]</span>,</span>
<span class="r-in">      blankPlotPos<span class="op">=</span><span class="fl">22</span>,</span>
<span class="r-in">      displayMAD<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in">   <span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in">   <span class="va">ggdf</span> <span class="op">&lt;-</span> <span class="fu">ggjammaplot</span><span class="op">(</span><span class="va">GeneSE</span>,</span>
<span class="r-in">      assay_name<span class="op">=</span><span class="st">"counts"</span>,</span>
<span class="r-in">      whichSamples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span>,</span>
<span class="r-in">      return_type<span class="op">=</span><span class="st">"data"</span>,</span>
<span class="r-in">      titleBoxColor<span class="op">=</span><span class="va">titleBoxColor</span><span class="op">)</span></span>
<span class="r-in">   <span class="va">highlightPoints1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jamba/man/tcount.html" class="external-link">tcount</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ggdf</span>, <span class="va">mean</span> <span class="op">&gt;</span> <span class="fl">15</span> <span class="op">&amp;</span> <span class="va">difference</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">$</span><span class="va">item</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in">   <span class="va">highlightPoints2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ggdf</span>, <span class="va">name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"CA1CB492"</span> <span class="op">&amp;</span></span>
<span class="r-in">      <span class="va">difference</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">4.5</span><span class="op">)</span><span class="op">$</span><span class="va">item</span>;</span>
<span class="r-in">   <span class="va">highlightPoints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span class="r-in">      divergent<span class="op">=</span><span class="va">highlightPoints1</span>,</span>
<span class="r-in">      low_CA1CB492<span class="op">=</span><span class="va">highlightPoints2</span><span class="op">)</span>;</span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">ggdf_h</span> <span class="op">&lt;-</span> <span class="fu">ggjammaplot</span><span class="op">(</span><span class="va">GeneSE</span>,</span>
<span class="r-in">      assay_name<span class="op">=</span><span class="st">"counts"</span>,</span>
<span class="r-in">      highlightPoints<span class="op">=</span><span class="va">highlightPoints</span>,</span>
<span class="r-in">      whichSamples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span>,</span>
<span class="r-in">      return_type<span class="op">=</span><span class="st">"data"</span>,</span>
<span class="r-in">      titleBoxColor<span class="op">=</span><span class="va">titleBoxColor</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in">   <span class="co"># you can use output from `jammaplot()` as input to `ggjammaplot()`:</span></span>
<span class="r-in">   <span class="va">jp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="jammaplot.html">jammaplot</a></span><span class="op">(</span><span class="va">GeneSE</span>,</span>
<span class="r-in">      outlierMAD<span class="op">=</span><span class="fl">2</span>,</span>
<span class="r-in">      doPlot<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span class="r-in">      assay_name<span class="op">=</span><span class="st">"raw_counts"</span>,</span>
<span class="r-in">      filterFloor<span class="op">=</span><span class="fl">1e-10</span>,</span>
<span class="r-in">      filterFloorReplacement<span class="op">=</span><span class="cn">NA</span>,</span>
<span class="r-in">      centerGroups<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">GeneSE</span><span class="op">)</span><span class="op">$</span><span class="va">Compartment</span>,</span>
<span class="r-in">      subtitleBoxColor<span class="op">=</span><span class="fu">farrisdata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/farrisdata/man/colorSub.html" class="external-link">colorSub</a></span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">GeneSE</span><span class="op">)</span><span class="op">$</span><span class="va">Compartment</span><span class="op">)</span><span class="op">]</span>,</span>
<span class="r-in">      useRank<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>;</span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">gg1</span> <span class="op">&lt;-</span> <span class="fu">ggjammaplot</span><span class="op">(</span><span class="va">jp2</span>,</span>
<span class="r-in">      ncol<span class="op">=</span><span class="fl">6</span>,</span>
<span class="r-in">      titleBoxColor<span class="op">=</span><span class="va">titleBoxColor</span><span class="op">)</span>;</span>
<span class="r-in">   <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gg1</span><span class="op">)</span>;</span>
<span class="r-in">   <span class="op">}</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>package ‘SummarizedExperiment’ was built under R version 3.6.2</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>package ‘S4Vectors’ was built under R version 3.6.3</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>package ‘IRanges’ was built under R version 3.6.2</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>package ‘GenomeInfoDb’ was built under R version 3.6.3</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>package ‘DelayedArray’ was built under R version 3.6.3</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>package ‘matrixStats’ was built under R version 3.6.2</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>package ‘BiocParallel’ was built under R version 3.6.2</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: stat</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 534 rows containing non-finite values (stat_density2d).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 7776 rows containing missing values (geom_raster).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: stat</span>
<span class="r-plt img"><img src="ggjammaplot-1.png" alt="" width="700" height="433"></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 140 rows containing non-finite values (stat_density2d).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 7776 rows containing missing values (geom_raster).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: stat</span>
<span class="r-plt img"><img src="ggjammaplot-2.png" alt="" width="700" height="433"></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 322 rows containing non-finite values (stat_density2d).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 7776 rows containing missing values (geom_raster).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: stat</span>
<span class="r-plt img"><img src="ggjammaplot-3.png" alt="" width="700" height="433"></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 317 rows containing non-finite values (stat_density2d).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 7636 rows containing missing values (geom_raster).</span>
<span class="r-plt img"><img src="ggjammaplot-4.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by James Ward.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

